### ISSUE TITLE
5.10.1 perlretut section "A bit of magic: executing Perl code in a regular expression" documentation

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#9917]
Migrated from [rt.perl.org#69903](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=69903) (status was 'resolved')

Searchable as RT69903$

**Comment by p5pRT** [comment]
### From casiano@ull.es

#### Created by casiano@ull.es



The "perlretut" doc\. in section 
"A bit of magic&#8203;: executing Perl code in a regular expression"
says&#8203;:

\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.
At first glance\, you'd think that it shouldn't print\, because obviously
the ddd  isn't going to match the target string\. But look at this example&#8203;:

&nbsp;   1\. $x =~ /abc\(?\{print "Hi Mom\!";\}\)\[d\]dd/; \# doesn't match\,
&nbsp;   2\. \# but \_does\_ print

Hmm\. What happened here? If you've been following along\, you know that
the above pattern should be effectively the same as the last one \-\-
enclosing the d in a character class isn't going to change what it
matches\. So why does the first not print while the second one does?
\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.\.

See the following code&#8203;:

pl@&#8203;nereida&#8203;:\~/Lperltesting$ cat embededcode\.pl
$x = "abcdef";
$x =~ /abc\(?\{print "Hi Mom\!";\}\)\[d\]dd/;

When executed with 5\.8\.8 has the documented behavior&#8203;:
pl@&#8203;nereida&#8203;:\~/Lperltesting$ perl5\.8\.8 embededcode\.pl

This is not the case in 5\.10\.1&#8203;:

Hi Mom\!pl@&#8203;nereida&#8203;:\~/Lperltesting$ perl5\.10\.1 embededcode\.pl
pl@&#8203;nereida&#8203;:\~/Lperltesting$

It looks as if the documentation has been left behind 
the implementation\.

<details><summary>Perl Info</summary>

```

Flags:
    category=docs
    severity=low

Site configuration information for perl 5.10.1:

Configured by root at Mon Aug 10 09:51:09 WEST 2009.

Summary of my perl5 (revision 5 version 10 subversion 1) configuration:
   
  Platform:
    osname=linux, osvers=2.6.24-24-generic, archname=x86_64-linux
    uname='linux europa 2.6.24-24-generic #1 smp tue jul 7 19:10:36 utc 2009 x86_64 gnulinux '
    config_args='-des -Dprefix=/usr/local/lib/perl/5.10.1'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2',
    cppflags='-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.2.4 (Ubuntu 4.2.4-1ubuntu4)', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib /lib64 /usr/lib64
    libs=-lnsl -ldl -lm -lcrypt -lutil -lc
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lc
    libc=/lib/libc-2.7.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.7'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -L/usr/local/lib -fstack-protector'

Locally applied patches:
    RC1
    
    PERL_GIT_UNPUSHED_COMMITS    	/* do not remove this line */
    PERL_GIT_UNCOMMITTED_CHANGES	/* do not remove this line */


@INC for perl 5.10.1:
    /usr/local/lib/perl/5.10.1/lib/5.10.1/x86_64-linux
    /usr/local/lib/perl/5.10.1/lib/5.10.1
    /usr/local/lib/perl/5.10.1/lib/site_perl/5.10.1/x86_64-linux
    /usr/local/lib/perl/5.10.1/lib/site_perl/5.10.1
    .


Environment for perl 5.10.1:
    HOME=/home/pl
    LANG=es_ES.UTF-8
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/pl/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/lib/perl/5.10.1/bin/
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @davidnicol

> It looks as if the documentation has been left behind
> the implementation\.

Is the documented behavior tested for? If not it should be

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @ikegami

On Tue\, Oct 20\, 2009 at 4&#8203;:48 PM\, David Nicol \<davidnicol@&#8203;gmail\.com> wrote&#8203;:

> > It looks as if the documentation has been left behind
> > the implementation\.
>
> Is the documented behavior tested for? If not it should be
>

The example is demonstrating that the order in which the components of the
regex pattern are matched is undefined\. It's not possible to test that the
matching order is undefined \(i\.e\. the documented behaviour\)\.

We could test if the specific output is produced in order to know when
perlretut needs to be updated\, but we shouldn't commit ourselves to the
behaviour in the example\.

Something like

\{
&nbsp;   my $f = 3;
&nbsp;   'abcdef' =~ /abc\(?\{ $f=4 \}\)ddd/;
&nbsp;   is\($f\, 3\, "perlretut example\, part 1"\)
&nbsp;       or diag\("This test and perlreftut needs to be updated"\);
\}

\{
&nbsp;   my $f = 3;
&nbsp;   'abcdef' =~ /abc\(?\{ $f=4 \}\)\[d\]dd/;
&nbsp;   is\($f\, 4\, "perlretut example\, part 2"\)
&nbsp;       or diag\("This test and perlreftut needs to be updated"\);
\}

**Comment by p5pRT** [comment]
### From @davidnicol

On Tue\, Oct 20\, 2009 at 5&#8203;:18 PM\, Eric Brine \<ikegami@&#8203;adaelis\.com> wrote&#8203;:
>>
>> Is the documented behavior tested for? If not it should be
>
> The example is demonstrating that the order in which the components of the
> regex pattern are matched is undefined\. It's not possible to test that the
> matching order is undefined \(i\.e\. the documented behaviour\)\.

Having the example in there will make people think that it's supposed
to work that way\. But until reading EB's response\, I thought regex
parts were defined to match right to left\, greedy\, then backtrack\,
instead of that simply being the current implementation of an
undefined behavior\.

If we upgrade this behavior from undependable implementation detail to
language definition by testing it\, it becomes a bug\.

See for instance the dismissals of any non\-bug concerning long\-running
regexen that are quick\-running if done backwards; google for
"sexeger\."

Why doesn't the executable piece run any more? To do the documentation
right\, this would be a place where footnotes about behavior changes
between versions would be good\.

**Comment by p5pRT** [comment]
### From alex@chmrr.net

At Tue Oct 20 03&#8203;:56&#8203;:14 \-0400 2009\, casiano@&#8203;ull\.es \(via RT\) wrote&#8203;:
> pl@&#8203;nereida&#8203;:\~/Lperltesting$ cat embededcode\.pl
> $x = "abcdef";
> $x =~ /abc\(?\{print "Hi Mom\!";\}\)\[d\]dd/;

This is because perl got smarter&#8203;:

commit 7f6f358c56bfa8125e150dbfb3ff2d549c2768cd
Author&#8203;: Yves Orton \<demerphq@&#8203;gmail\.com>
Date&#8203;:   Sun May 21 01&#8203;:16&#8203;:33 2006 \+0200

&nbsp;    optimize /\[x\]/ to /x/\.
&nbsp;    Message\-Id&#8203;:  \<9b18b3110605201416i3dc44c39k86cd9b2f47f5cda0@&#8203;mail\.gmail\.com>
&nbsp;    
&nbsp;    p4raw\-id&#8203;: //depot/perl@&#8203;28262

The documentation could be "fixed" by replacing the regexp with&#8203;:

&nbsp;    $x =~ /abc\(?\{print "Hi Mom\!";\}\)\(?=ddd\)/; \# doesn't match\,
&nbsp;                                             \# but \_does\_ print

\.\.\.but in some sense that example is us waging a \(hopefully\) losing
war against the regex optimizer\.
 \- Alex
\-\- 
Networking \-\- only one letter away from not working

**Comment by p5pRT** [comment]
@rgs - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @rgarcia

2009/10/21 Alex Vandiver \<chmrr@&#8203;chmrr\.net>&#8203;:
> At Tue Oct 20 03&#8203;:56&#8203;:14 \-0400 2009\, casiano@&#8203;ull\.es \(via RT\) wrote&#8203;:
>> pl@&#8203;nereida&#8203;:\~/Lperltesting$ cat embededcode\.pl
>> $x = "abcdef";
>> $x =~ /abc\(?\{print "Hi Mom\!";\}\)\[d\]dd/;
>
> This is because perl got smarter&#8203;:
>
> commit 7f6f358c56bfa8125e150dbfb3ff2d549c2768cd
> Author&#8203;: Yves Orton \<demerphq@&#8203;gmail\.com>
> Date&#8203;:   Sun May 21 01&#8203;:16&#8203;:33 2006 \+0200
>
>    optimize /\[x\]/ to /x/\.
>    Message\-Id&#8203;:  \<9b18b3110605201416i3dc44c39k86cd9b2f47f5cda0@&#8203;mail\.gmail\.com>
>
>    p4raw\-id&#8203;: //depot/perl@&#8203;28262
>
> The documentation could be "fixed" by replacing the regexp with&#8203;:
>
>    $x =~ /abc\(?\{print "Hi Mom\!";\}\)\(?=ddd\)/; \# doesn't match\,
>                                             \# but \_does\_ print
>
> \.\.\.but in some sense that example is us waging a \(hopefully\) losing
> war against the regex optimizer\.

I applied this :

commit 871673162d98aa63f3298e094e8afb43bc4230df
Author&#8203;: Rafael Garcia\-Suarez \<rgs@&#8203;consttype\.org>
Date&#8203;:   Fri Oct 23 00&#8203;:22&#8203;:16 2009 \+0200

&nbsp;    \[perl \#69903\] 5\.10\.1 perlretut section "A bit of magic&#8203;: executing
Perl code in a regular expression" documentation

&nbsp;    The docs were now wrong\, due to new optimisations to the regexp engine\.

<details><summary>Inline Patch</summary>

```diff
diff --git a/pod/perlretut.pod b/pod/perlretut.pod
index 22fc44a..6c5c2e9 100644
--- a/pod/perlretut.pod
+++ b/pod/perlretut.pod
@@ -2474,11 +2474,11 @@ At first glance, you'd think that it shouldn't

```
</details>
print, because obviously
 the C\<ddd> isn't going to match the target string\. But look at this
 example&#8203;:

\-    $x =~ /abc\(?\{print "Hi Mom\!";\}\)\[d\]dd/; \# doesn't match\,
\-                                           \# but \_does\_ print
\+    $x =~ /abc\(?\{print "Hi Mom\!";\}\)\[dD\]dd/; \# doesn't match\,
\+                                            \# but \_does\_ print

 Hmm\. What happened here? If you've been following along\, you know that
\-the above pattern should be effectively the same as the last one \-\-
\+the above pattern should be effectively \(almost\) the same as the last one \-\-
 enclosing the d in a character class isn't going to change what it
 matches\. So why does the first not print while the second one does?

@&#8203;@&#8203; \-2487\,7 \+2487\,7 @&#8203;@&#8203; case\, all the engine sees are plain old
characters \(aside from the
 C\<?\{\}> construct\)\. It's smart enough to realize that the string 'ddd'
 doesn't occur in our target string before actually running the pattern
 through\. But in the second case\, we've tricked it into thinking that our
\-pattern is more complicated than it is\. It takes a look\, sees our
\+pattern is more complicated\. It takes a look\, sees our
 character class\, and decides that it will have to actually run the
 pattern to determine whether or not it matches\, and in the process of
 running it hits the print statement before it discovers that we don't

**Comment by p5pRT** [comment]
### From @demerphq

2009/10/21 Alex Vandiver \<chmrr@&#8203;chmrr\.net>&#8203;:
> At Tue Oct 20 03&#8203;:56&#8203;:14 \-0400 2009\, casiano@&#8203;ull\.es \(via RT\) wrote&#8203;:
>> pl@&#8203;nereida&#8203;:\~/Lperltesting$ cat embededcode\.pl
>> $x = "abcdef";
>> $x =~ /abc\(?\{print "Hi Mom\!";\}\)\[d\]dd/;
>
> This is because perl got smarter&#8203;:
>
> commit 7f6f358c56bfa8125e150dbfb3ff2d549c2768cd
> Author&#8203;: Yves Orton \<demerphq@&#8203;gmail\.com>
> Date&#8203;:   Sun May 21 01&#8203;:16&#8203;:33 2006 \+0200
>
>    optimize /\[x\]/ to /x/\.
>    Message\-Id&#8203;:  \<9b18b3110605201416i3dc44c39k86cd9b2f47f5cda0@&#8203;mail\.gmail\.com>
>
>    p4raw\-id&#8203;: //depot/perl@&#8203;28262
>
> The documentation could be "fixed" by replacing the regexp with&#8203;:
>
>    $x =~ /abc\(?\{print "Hi Mom\!";\}\)\(?=ddd\)/; \# doesn't match\,
>                                             \# but \_does\_ print
>
> \.\.\.but in some sense that example is us waging a \(hopefully\) losing
> war against the regex optimizer\.

Right\. Put two characters into the charclass and the optimisation wont kick in\.

And the reason it doesnt match once that optimisation kicks in \(or the
equivalent code is used\) is that the optimiser gets two "fixed"
strings to look for "abc" and "ddd"\,  and because the "ddd" is later
in the pattern it will choose that\. Once a fixed string in a pattern
has been identified the regex engine will search for that string first
and if it is not found not bother with trying the pattern any further\.

I believe that \(??\{ \.\.\. \} is special cased to bypass this\, but im not 100% sure\.

Anyway\, this is not a bug\.

Yves



\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
### From @demerphq

2009/10/21 David Nicol \<davidnicol@&#8203;gmail\.com>&#8203;:
> On Tue\, Oct 20\, 2009 at 5&#8203;:18 PM\, Eric Brine \<ikegami@&#8203;adaelis\.com> wrote&#8203;:
>>>
>>> Is the documented behavior tested for? If not it should be
>>
>> The example is demonstrating that the order in which the components of the
>> regex pattern are matched is undefined\. It's not possible to test that the
>> matching order is undefined \(i\.e\. the documented behaviour\)\.
>
> Having the example in there will make people think that it's supposed
> to work that way\. But until reading EB's response\, I thought regex
> parts were defined to match right to left\, greedy\, then backtrack\,
> instead of that simply being the current implementation of an
> undefined behavior\.

You are discussing what happens in a successful match\. However this
report is about a \*failing\* match\.

Perl has never made any promises about what happens in a failing match\.

> If we upgrade this behavior from undependable implementation detail to
> language definition by testing it\, it becomes a bug\.
>
> See for instance the dismissals of any non\-bug concerning long\-running
> regexen that are quick\-running if done backwards; google for
> "sexeger\."
>
> Why doesn't the executable piece run any more? To do the documentation
> right\, this would be a place where footnotes about behavior changes
> between versions would be good\.

Because the optimizer is now smart enough to detect that /\[d\]dd/ and
/ddd/ are the same pattern and replaces the former with the latter as
it is more efficient\. This in turn allows another optimisation to kick
in which uses the fact there must be a "ddd" in the string to decide
that this string cannot match and thus there is no point entering the
costly regex engine internals\.

Additionally once the "ddd" is found it knows that any successful
match in this case must start 3 chars to the left of the "ddd"\. So for
instance "abcabcabcabcddd" will only trigger the code sub once\, not 4
times\. So we fail fast\, and prune the search space with this
optimization\.

cheers\,
Yves



\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"
